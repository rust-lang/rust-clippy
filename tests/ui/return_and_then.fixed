#![warn(clippy::return_and_then)]

fn main() {
    fn test_opt_block(opt: Option<i32>) -> Option<i32> {
        let n = opt?;
        let mut ret = n + 1;
                    ret += n;
                    if n > 1 { Some(ret) } else { None }
    }

    fn test_opt_func(opt: Option<i32>) -> Option<i32> {
        let n = opt?;
        test_opt_block(Some(n))
    }

    fn test_call_chain() -> Option<i32> {
        let n = gen_option(1)?;
        test_opt_func(Some(n))
    }

    fn test_res_block(opt: Result<i32, i32>) -> Result<i32, i32> {
        let n = opt?;
        if n > 1 { Ok(n + 1) } else { Err(n) }
    }

    fn test_res_func(opt: Result<i32, i32>) -> Result<i32, i32> {
        let n = opt?;
        test_res_block(Ok(n))
    }
}

fn gen_option(n: i32) -> Option<i32> {
    Some(n)
}
