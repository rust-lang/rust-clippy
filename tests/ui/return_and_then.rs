#![warn(clippy::return_and_then)]

fn main() {
    fn test_opt_block(opt: Option<i32>) -> Option<i32> {
        opt.and_then(|n| {
            let mut ret = n + 1;
            ret += n;
            if n > 1 { Some(ret) } else { None }
        })
    }

    fn test_opt_func(opt: Option<i32>) -> Option<i32> {
        opt.and_then(|n| test_opt_block(Some(n)))
    }

    fn test_call_chain() -> Option<i32> {
        gen_option(1).and_then(|n| test_opt_func(Some(n)))
    }

    fn test_res_block(opt: Result<i32, i32>) -> Result<i32, i32> {
        opt.and_then(|n| if n > 1 { Ok(n + 1) } else { Err(n) })
    }

    fn test_res_func(opt: Result<i32, i32>) -> Result<i32, i32> {
        opt.and_then(|n| test_res_block(Ok(n)))
    }
}

fn gen_option(n: i32) -> Option<i32> {
    Some(n)
}
