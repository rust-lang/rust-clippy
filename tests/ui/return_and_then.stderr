error: use the question mark operator instead of an `and_then` call
  --> tests/ui/return_and_then.rs:5:9
   |
LL | /         opt.and_then(|n| {
LL | |             let mut ret = n + 1;
LL | |             ret += n;
LL | |             if n > 1 { Some(ret) } else { None }
LL | |         })
   | |__________^
   |
   = note: `-D clippy::return-and-then` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::return_and_then)]`
help: try
   |
LL ~         let n = opt?;
LL +         let mut ret = n + 1;
LL +                     ret += n;
LL +                     if n > 1 { Some(ret) } else { None }
   |

error: use the question mark operator instead of an `and_then` call
  --> tests/ui/return_and_then.rs:13:9
   |
LL |         opt.and_then(|n| test_opt_block(Some(n)))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try
   |
LL ~         let n = opt?;
LL +         test_opt_block(Some(n))
   |

error: use the question mark operator instead of an `and_then` call
  --> tests/ui/return_and_then.rs:17:9
   |
LL |         gen_option(1).and_then(|n| test_opt_func(Some(n)))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try
   |
LL ~         let n = gen_option(1)?;
LL +         test_opt_func(Some(n))
   |

error: use the question mark operator instead of an `and_then` call
  --> tests/ui/return_and_then.rs:21:9
   |
LL |         opt.and_then(|n| if n > 1 { Ok(n + 1) } else { Err(n) })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try
   |
LL ~         let n = opt?;
LL +         if n > 1 { Ok(n + 1) } else { Err(n) }
   |

error: use the question mark operator instead of an `and_then` call
  --> tests/ui/return_and_then.rs:25:9
   |
LL |         opt.and_then(|n| test_res_block(Ok(n)))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try
   |
LL ~         let n = opt?;
LL +         test_res_block(Ok(n))
   |

error: aborting due to 5 previous errors

